{"cells": [{"cell_type": "markdown", "metadata": {}, "source": ["#####################################################################################################################<br>\n", "Python script to construct and run coexistence simulations of intrinsically disordered proteins using the HPS model[1] <br>\n", "[1] Sequence determinants of protein phase behavior from a coarse-grained model <br>\n", "    GL Dignon, W Zheng, YC Kim, RB Best, J Mittal - PLoS computational biology, 2018<br>\n", "<br>\n", "Authors: Roshan M Regy, Wenwei Zheng, Gregory L. Dignon, Jeetain Mittal <br>\n", "#####################################################################################################################<br>\n", "Prerequisities: This code was tested using HOOMD v2.9.0 with azplugins installed and GSD v2.2.0 <br>\n", "<br>\n", "Usage: python Simulation_with_HOOMD.py YOUR_FASTA_PROTEIN_SEQUENCE_FILE <br>\n", "<br>\n", "This code has the following steps:<br>\n", "<br>\n", "(1) Create a single chain initial configuration as a linear polymer chain --> generates start.gsd <br>\n", "<br>\n", "(2) Replicate the single chain initial configuration using the replicate() function in HOOMD to <br>\n", "  user defined (nx*ny*nz) system size and then reduce the dimensions of the replicated configuration <br>\n", "  to form a cubic box with user defined dimensions using hoomd.update.box_resize() while <br>\n", "  running an NVT simulation at user defined temperature and time step  --> generates resize.gsd, resize.dcd, resize.log<br>\n", "<br>\n", "(3) Unwrap molecules over the periodic boundary in z and extend the box in the z direction to create a slab configuration <br>\n", "  --> generates box2slab_extend.gsd <br>\n", "<br>\n", "(4) Run a production NVT simulation at user defined temperature and time step and for user defined number of steps<br>\n", "  --> generates production.dcd, restart_tmp1.gsd, restart_tmp2.gsd, production.log, stress.log  <br>\n", "#####################################################################################################################"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["import sys,os,numpy as np\n", "import hoomd, hoomd.md as md\n", "import hoomd.deprecated as old\n", "from hoomd import azplugins\n", "import gsd, gsd.hoomd, gsd.pygsd "]}, {"cell_type": "markdown", "metadata": {}, "source": ["SYSTEM SIZE"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["nx=5 # Replicate system in x direction \n", "ny=5 # Replicate system in y direction \n", "nz=4 # Replicate system in z direction --> Total system size would be nx*ny*nz\n", "# Please keep magnitudes of nx, ny and nz as close to each other \n", "# as possible to avoid large replicated system box sizes in a specific direction "]}, {"cell_type": "markdown", "metadata": {}, "source": ["SLAB DIMENSIONS"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["boxsize=15.0 # The x and y dimensions of the slab configuration in nanometers\n", "slab_z_length = 280.0 # The z dimension of the slab configuration in nanometers"]}, {"cell_type": "markdown", "metadata": {}, "source": ["RESIZING RUN PARAMETERS"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["resize_T=300 # Temperature for resizing run in Kelvin\n", "resize_steps=600#500000 # Number of steps (>500) to be used for resizing the box to boxsize set previously\n", "if resize_steps<500:\n", "    resize_steps=501\n", "resize_dt=0.01 # Time step in picoseconds for box resizing "]}, {"cell_type": "markdown", "metadata": {}, "source": ["PRODUCTION RUN PARAMETERS"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["production_dt=0.01 # Time step for production run in picoseconds\n", "production_steps=100000000 # Total number of steps \n", "production_T=300 # Temperature for production run in Kelvin"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["seq={'R':'ARG','H':'HIS','K':'LYS','D':'ASP','E':'GLU',\n", "     'S':'SER','T':'THR','N':'ASN','Q':'GLN','C':'CYS',\n", "     'U':'SEC','G':'GLY','P':'PRO','A':'ALA','V':'VAL',\n", "     'I':'ILE','L':'LEU','M':'MET','F':'PHE','Y':'TYR',\n", "     'W':'TRP'}"]}, {"cell_type": "markdown", "metadata": {}, "source": ["##### Read one letter amino acid sequence from file"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["filein = sys.argv[1]\n", "fileout='%s_seq3.dat'%(filein)\n", "nline=1\n", "count=0\n", "fout=open(fileout,'w')\n", "with open(filein,'r') as fid:\n", "    for i in fid:\n", "        if i[0]!='#':\n", "            for j in i:\n", "                if j in seq:\n", "                    fout.write(' %s'%seq[j])\n", "                    count+=1\n", "                    if count==nline:\n", "                        fout.write('\\n')\n", "                        count=0\n", "fout.close()"]}, {"cell_type": "markdown", "metadata": {}, "source": ["#### 1.2 Read sequence and force field parameters<br>\n", "##### Input parameters for all the amino acids (force field)"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["ff_para = 'stats_module.dat'\n", "aalist={}\n", "with open(ff_para,'r') as fid:\n", "    for i in fid:\n", "        if i[0]!='#':\n", "            tmp=i.rsplit()\n", "            aalist[tmp[0]]=np.loadtxt(tmp[1:],dtype=float)\n", "aakeys=list(aalist.keys())\n", "# This translates each amino acid type into a number, which will be used in HOOMD\n", "# For example, GLY is with an ID of 10\n", "aamass=[]\n", "aacharge=[]\n", "aaradius=[]\n", "aahps=[]\n", "print ('aakeys')\n", "print (aakeys)\n", "print ('aalist[i][1]')\n", "print (aalist[aakeys[1]][1])\n", "for i in aakeys:\n", "    aamass.append(aalist[i][0])\n", "    aacharge.append(aalist[i][1])\n", "    aaradius.append(aalist[i][2])\n", "    aahps.append(aalist[i][3])\n", "# Now we can translate the entire sequence into a number code according to the order in 'aakeys'\n", "chain_id=[]\n", "chain_mass=[]\n", "chain_charge=[]\n", "with open(fileout,'r') as fid:\n", "    for i in fid:\n", "        iname=i.rsplit()[0]\n", "        chain_id.append(aakeys.index(iname))\n", "        chain_mass.append(aalist[iname][0])\n", "        chain_charge.append(aalist[iname][1])"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["bond_length=0.38\n", "chain_length=len(chain_id)\n", "box_length=bond_length*chain_length+10\n", "# #### 1.3 Now we can build HOOMD data structure for one single frame\n", "s=gsd.hoomd.Snapshot()\n", "s.particles.N = chain_length\n", "s.particles.types = aakeys\n", "s.particles.typeid = chain_id\n", "s.particles.mass = chain_mass\n", "s.particles.charge = chain_charge"]}, {"cell_type": "markdown", "metadata": {}, "source": [" Build initial position as a linear chain"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["pos=[]\n", "for i in range(len(chain_id)):\n", "    # Change the z-coordinate to have a linear chain\n", "    pos.append((0,0,(i-int(len(chain_id)/2))*bond_length))\n", "pos=np.array(pos)\n", "s.particles.position= pos"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Initialize bond"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["nbonds=chain_length-1\n", "s.bonds.N = nbonds\n", "s.bonds.types = ['AA_bond']\n", "s.bonds.typeid = [0]*(nbonds)\n", "bond_pairs=np.zeros((nbonds,2),dtype=int)\n", "for i in range(nbonds):\n", "    bond_pairs[i,:]=np.array([i,i+1])\n", "s.bonds.group = bond_pairs"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Box size"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["s.configuration.dimensions=3\n", "s.configuration.box=[box_length,box_length,box_length,0,0,0]\n", "s.configuration.step=0"]}, {"cell_type": "markdown", "metadata": {}, "source": ["#### 1.4 Write intial singe chain gsd file"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["f = gsd.hoomd.open(name='start.gsd', mode='wb')\n", "f.append(s)\n", "f.close()\n", "#################################################################################################\n", "# ## start.gsd contains one single chain of the given protein \n", "# ----------------------------------------------------------------------------------------------\n", "# ## 2.0 Replicate single chain system to given number of chains for a slab \n", "################################################################################################\n", "hoomd.context.initialize()\n", "system = hoomd.init.read_gsd('start.gsd')"]}, {"cell_type": "markdown", "metadata": {}, "source": ["### Replicate the single chain here. Remember total number of chains = nx*ny*nz"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["system.replicate(nx=nx,ny=ny,nz=nz) "]}, {"cell_type": "markdown", "metadata": {}, "source": ["#### Bonds"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["harmonic=hoomd.md.bond.harmonic()\n", "harmonic.bond_coeff.set('AA_bond',k=8368,r0=bond_length)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["#### Neighborlist and exclusions"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["nl = hoomd.md.nlist.cell()\n", "nl.reset_exclusions(exclusions=['1-2', 'body'])"]}, {"cell_type": "markdown", "metadata": {}, "source": ["#### Pairwise interactions<br>\n", "b = hoomd.md.pair.lj_lambda(r_cut=0, nlist=nl) "]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["nb = azplugins.pair.ashbaugh(r_cut=0, nlist=nl)\n", "for i in aakeys:\n", "    for j in aakeys:\n", "        nb.pair_coeff.set(i,j,lam=(aalist[i][3]+aalist[j][3])/2.,\n", "                          epsilon=0.8368, sigma=(aalist[i][2]+aalist[j][2])/10./2.,r_cut=2.0)    \n", "        \n", "# #### Electrostatics\n", "yukawa = hoomd.md.pair.yukawa(r_cut=0.0, nlist=nl)\n", "for i,atom1 in enumerate(aakeys):\n", "    for j,atom2 in enumerate(aakeys):\n", "        yukawa.pair_coeff.set(atom1,atom2,epsilon=aalist[atom1][1]*aalist[atom2][1]*1.73136, kappa=1.0, r_cut=3.5) \n", "# #### Group Particles\n", "all = hoomd.group.all()"]}, {"cell_type": "markdown", "metadata": {}, "source": ["#### Set up integrator"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["hoomd.md.integrate.mode_standard(dt=resize_dt) # Time units in ps\n", "kTinput=resize_T * 8.3144598/1000.\n", "integrator = hoomd.md.integrate.langevin(group=all, kT=kTinput, seed=63535)\n", "# #### Resize the box after replication to 15x15x15nm  \n", "hoomd.update.box_resize(L=hoomd.variant.linear_interp([(0,system.box.Lx),(resize_steps-500,boxsize)]),scale_particles=True)\n", "for cnt,i in enumerate(aakeys):\n", "    integrator.set_gamma(i,gamma=aamass[cnt]/1000.0)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["#### Output log file with box dimensions and restart file after box resizing"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["hoomd.analyze.log(filename='resize.log', quantities=['potential_energy','kinetic_energy','temperature','pressure_xx','pressure_yy','pressure_zz','lx','ly','lz'], period=10000, overwrite=True, header_prefix='#')\n", "hoomd.dump.gsd('resize.gsd', period=10000, group=all, truncate=True)\n", "hoomd.dump.dcd('resize.dcd', period=10000, group=all, overwrite=True)\n", "# #### Run resizing simulation\n", "hoomd.run(tsteps=resize_steps)\n", "#################################################################################################\n", "# ## resize.gsd contains replicated system with 100 chains and box size of 15x15x15 nm \n", "# ----------------------------------------------------------------------------------------------\n", "# ## 3.0 Extend 15nm cube to 15x15x280nm slab \n", "################################################################################################\n", "def extend(s):\n", "    boxdim = s.configuration.box[:3]\n", "    zmin,zmax,dz = -boxdim[2]/2., boxdim[2]/2., boxdim[2]\n", "    pos1 =  s.particles.position\n", "    pos = pos1.copy()\n", "    skip=0\n", "    ncomp=1\n", "    for k in range(ncomp):\n", "        nchain = int(s.particles.N/chain_length)\n", "        nres = chain_length\n", "        for i in range(nchain):\n", "            mol_coord = pos[i*nres+skip:(i+1)*nres+skip,2]\n", "            for j in range(1,nres):\n", "                dist2 = (mol_coord[j] - mol_coord[j-1])**2\n", "                if dist2 > 8:\n", "                    excess = np.sign(mol_coord[j] - mol_coord[j-1])*dz\n", "                    mol_coord[j] = mol_coord[j] - excess \n", "                com = np.mean(mol_coord)\n", "                if com < zmin:\n", "                    mol_coord += dz\n", "                elif com > zmax:\n", "                    mol_coord -= dz\n", "            pos[i*nres+skip:(i+1)*nres+skip,2] = mol_coord\n", "        skip += nchain*nres\n", "    return pos"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["f = gsd.pygsd.GSDFile(open('resize.gsd','rb'))\n", "t = gsd.hoomd.HOOMDTrajectory(f)\n", "s1 = t[0]\n", "s = gsd.hoomd.Snapshot()\n", "s.particles.N = s1.particles.N\n", "s.particles.types = s1.particles.types \n", "s.particles.typeid = s1.particles.typeid \n", "s.particles.mass = s1.particles.mass\n", "s.particles.charge = s1.particles.charge\n", "s.particles.position = extend(s1)\n", "s.bonds.N = s1.bonds.N\n", "s.bonds.types = s1.bonds.types\n", "s.bonds.typeid = s1.bonds.typeid\n", "s.bonds.group = s1.bonds.group\n", "s.configuration.box = s1.configuration.box\n", "s.configuration.dimensions=3\n", "s.configuration.box = [s1.configuration.box[0],s1.configuration.box[1],slab_z_length,0,0,0] \n", "s.configuration.step = 0\n", "outfile = gsd.hoomd.open('box2slab_extend.gsd','wb')\n", "outfile.append(s)\n", "outfile.close()\n", "#################################################################################################\n", "# ### Minimized slab formed and saved in minimize.gsd\n", "#---------------------------------------------------------------------------------------------\n", "# ## 4.0. Run a production slab simulation using minimize.gsd from previous step\n", "################################################################################################\n", "hoomd.context.initialize()\n", "system = hoomd.init.read_gsd('box2slab_extend.gsd')"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["n_steps = production_steps # 1 microseconds"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["fileroot = 'Production'\n", "nl = hoomd.md.nlist.cell()"]}, {"cell_type": "markdown", "metadata": {}, "source": [" Bonds"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["harmonic = hoomd.md.bond.harmonic()\n", "harmonic.bond_coeff.set('AA_bond', k=8360, r0=0.381)\n", "## Nonbonded\n", "nl.reset_exclusions(exclusions=['1-2', 'body'])\n", "nb = azplugins.pair.ashbaugh(r_cut=0, nlist=nl)\n", "for i in aakeys:\n", "    for j in aakeys:\n", "        nb.pair_coeff.set(i,j,lam=(aalist[i][3]+aalist[j][3])/2.,\n", "                          epsilon=0.8368, sigma=(aalist[i][2]+aalist[j][2])/10./2.,r_cut=2.0)    "]}, {"cell_type": "markdown", "metadata": {}, "source": [" Electrostatics"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["yukawa = hoomd.md.pair.yukawa(r_cut=0.0, nlist=nl)\n", "for i,atom1 in enumerate(aakeys):\n", "    for j,atom2 in enumerate(aakeys):\n", "        yukawa.pair_coeff.set(atom1,atom2,epsilon=aalist[atom1][1]*aalist[atom2][1]*1.73136, kappa=1.0, r_cut=3.5) "]}, {"cell_type": "markdown", "metadata": {}, "source": [" Group Particles"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["all = hoomd.group.all()"]}, {"cell_type": "markdown", "metadata": {}, "source": [" Set up integrator"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["hoomd.md.integrate.mode_standard(dt=production_dt) # Time units in ps\n", "temp = production_T*0.00831446\n", "integrator = hoomd.md.integrate.langevin(group=all, kT=temp, seed=399991) # Temp is kT/0.00831446\n", "for cnt,i in enumerate(aakeys):\n", "    integrator.set_gamma(i,gamma=aamass[cnt]/1000.0)\n", "## Outputs\n", "hoomd.analyze.log(filename=fileroot+'.log', quantities=['potential_energy', 'pressure_xx', 'pressure_yy', 'pressure_zz', 'temperature','lx','ly','lz'], period=100000, overwrite=False, header_prefix='#')\n", "hoomd.analyze.log(filename='stress.log', quantities=['pressure_xy', 'pressure_xz', 'pressure_yz'], period=100000, overwrite=False, header_prefix='#') # Output stress tensor\n", "hoomd.dump.gsd('restart_tmp1.gsd', period=1000000, group=all, truncate=True)\n", "hoomd.dump.gsd('restart_tmp2.gsd', period=1000000, group=all, truncate=True, phase=500000)\n", "hoomd.dump.dcd(fileroot+'_dump.dcd', period=100000, group=all, overwrite=False)"]}, {"cell_type": "markdown", "metadata": {}, "source": [" Run simulation"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["hoomd.run_upto(production_steps, limit_hours=48)\n", "########################################################################################################"]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.4"}}, "nbformat": 4, "nbformat_minor": 2}